dht_options.go

```diff
@@ -0,0 +1,262 @@
+package dht
+
+import (
+	"fmt"
+	"time"
+
+	ds "github.com/ipfs/go-datastore"
+	dssync "github.com/ipfs/go-datastore/sync"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-record"
+)
+
+// ModeOpt describes what mode the dht should operate in
+type ModeOpt int
+
+const (
+	// ModeAuto utilizes EvtLocalReachabilityChanged events sent over the event bus to dynamically switch the DHT
+	// between Client and Server modes based on network conditions
+	ModeAuto ModeOpt = iota
+	// ModeClient operates the DHT as a client only, it cannot respond to incoming queries
+	ModeClient
+	// ModeServer operates the DHT as a server, it can both send and respond to queries
+	ModeServer
+)
+
+// Options is a structure containing all the options that can be used when constructing a DHT.
+type config struct {
+	datastore       ds.Batching
+	validator       record.Validator
+	mode            ModeOpt
+	protocols       []protocol.ID
+	bucketSize      int
+	disjointPaths   int
+	concurrency     int
+	maxRecordAge    time.Duration
+	enableProviders bool
+	enableValues    bool
+
+	routingTable struct {
+		refreshQueryTimeout time.Duration
+		refreshPeriod       time.Duration
+		autoRefresh         bool
+		latencyTolerance    time.Duration
+	}
+}
+
+// Apply applies the given options to this Option
+func (o *config) Apply(opts ...Option) error {
+	for i, opt := range opts {
+		if err := opt(o); err != nil {
+			return fmt.Errorf("dht option %d failed: %s", i, err)
+		}
+	}
+	return nil
+}
+
+// Option DHT option type.
+type Option func(*config) error
+
+// Defaults are the default DHT options. This option will be automatically
+// prepended to any options you pass to the DHT constructor.
+var Defaults = func(o *config) error {
```

Comment:
Should this be private for now?

Reply:
That's hard to do without either copying this inside the options package or breaking things. I'm just going to remove this.

Reply:
Nobody should be using it anyways.